### DISCUSSION TITLE
Fiwix

### DISCUSSION THREAD

**Discussion by toncho11** [jbruchon/elks#1447] [General]
He is an OS similar to ELKS https://github.com/mikaku/Fiwix
Fiwix does not run on 16 bit machines - only on 386 and above. It does not have networking, but it does run DOOM (lxdoom-1.4.4) using SVGAlib. Here is the source code of DOOM that could be eventually be adapted to run ELKS (on 386): https://sourceforge.net/projects/lxdoom/files/lxdoom/1.4.4/

**Comment by ghaerr** [comment]
Hello @toncho11,

Responding to this and the SVGALib thread on #1343: source code and operating systems that require 386+ CPUs to run also require a 32-bit C compiler, which ELKS doesn't have. I would have to look further into SVGALib, but IIRC, it also requires 32-bits, and on MSDOS requires DJGPP, a 32-bit compiler. I already have Doom supported in Microwindows, which would not require any changes or SVGALib porting, except that ELKS only supports the 16-bit version of Microwindows from circa 1999 or so.

The ELKS C compiler can support 32-bit pointers, but each pointer can only address a maximum of 64K data, and support for that requires changing all source code `char *` declarations to `char __far *`, since there is no large or huge model support. In addition, even after those changes, the program would have to be inspected and rewritten to ensure that any data referenced is smaller than 64K.

While it may be possible to get Doom to run in 16-bits, we'd want to start with a 16-bit version of Doom source. Other likely big problems are getting the framerate up to something fast enough to actually be ableto  have fun playing the game. On ELKS, reports are that even drawing a single full-screen 640x480 VGA screen takes possible an eighth to a quarter of a second on older hardware. (Run `nxworld` for timing results on your particular system to see actual details).

Of course, another option would be to somehow add DPMI support and 32-bit compiler support into ELKS, but both of those would be large projects.

Thank you!

**Comment by toncho11** [comment]
Thank you. 
I see it is going to be tough. A 16 bit Doom will always be very slow. 
DPMI you mean protected mode. 
Yeah, running on 16 bit machines is one of the main advantage of ELKS. DOOM can wait :)
